[RFC] Generate unicode tables with script

ZyX-I opened the pull request:
This commits removes large tables from mbyte.c and adds a script that generates them. Added lua script is a translation of unicode.vim tool (removed from the repository) to lua.

Downloading of file is done using `file(DOWNLOAD` when running CMake to generate build files and I am not entirely sure this is what should be done.

This results in the following update to the unicode tables:

``` Patch
--- /tmp/unicode_tables.old.generated.h 2015-03-15 20:13:21.438407881 +0300
+++ build/src/nvim/auto/unicode_tables.generated.h  2015-03-15 20:13:01.848017764 +0300
@@ -65,2 +65,3 @@ static const convertStruct toLower[] = {
   {0x376, 0x376, -1, 1},
+  {0x37f, 0x37f, -1, 116},
   {0x386, 0x386, -1, 38},
@@ -84,5 +85,6 @@ static const convertStruct toLower[] = {
   {0x4c1, 0x4cd, 2, 1},
-  {0x4d0, 0x524, 2, 1},
+  {0x4d0, 0x52e, 2, 1},
   {0x531, 0x556, 1, 48},
   {0x10a0, 0x10c5, 1, 7264},
+  {0x10c7, 0x10cd, 6, 7264},
   {0x1e00, 0x1e94, 2, 1},
@@ -134,5 +136,5 @@ static const convertStruct toLower[] = {
   {0x2ceb, 0x2ced, 2, 1},
-  {0xa640, 0xa65e, 2, 1},
-  {0xa662, 0xa66c, 2, 1},
-  {0xa680, 0xa696, 2, 1},
+  {0x2cf2, 0xa640, 31054, 1},
+  {0xa642, 0xa66c, 2, 1},
+  {0xa680, 0xa69a, 2, 1},
   {0xa722, 0xa72e, 2, 1},
@@ -143,4 +145,14 @@ static const convertStruct toLower[] = {
   {0xa78b, 0xa78b, -1, 1},
+  {0xa78d, 0xa78d, -1, -42280},
+  {0xa790, 0xa792, 2, 1},
+  {0xa796, 0xa7a8, 2, 1},
+  {0xa7aa, 0xa7aa, -1, -42308},
+  {0xa7ab, 0xa7ab, -1, -42319},
+  {0xa7ac, 0xa7ac, -1, -42315},
+  {0xa7ad, 0xa7ad, -1, -42305},
+  {0xa7b0, 0xa7b0, -1, -42258},
+  {0xa7b1, 0xa7b1, -1, -42282},
   {0xff21, 0xff3a, 1, 32},
   {0x10400, 0x10427, 1, 40},
+  {0x118a0, 0x118bf, 1, 32},
 };
@@ -199,4 +211,8 @@ static const convertStruct toUpper[] = {
   {0x25b, 0x25b, -1, -203},
+  {0x25c, 0x25c, -1, 42319},
   {0x260, 0x260, -1, -205},
+  {0x261, 0x261, -1, 42315},
   {0x263, 0x263, -1, -207},
+  {0x265, 0x265, -1, 42280},
+  {0x266, 0x266, -1, 42308},
   {0x268, 0x268, -1, -209},
@@ -204,2 +220,3 @@ static const convertStruct toUpper[] = {
   {0x26b, 0x26b, -1, 10743},
+  {0x26c, 0x26c, -1, 42305},
   {0x26f, 0x26f, -1, -211},
@@ -210,2 +227,3 @@ static const convertStruct toUpper[] = {
   {0x280, 0x283, 3, -218},
+  {0x287, 0x287, -1, 42282},
   {0x288, 0x288, -1, -218},
@@ -215,2 +233,3 @@ static const convertStruct toUpper[] = {
   {0x292, 0x292, -1, -219},
+  {0x29e, 0x29e, -1, 42258},
   {0x345, 0x345, -1, 84},
@@ -235,2 +254,3 @@ static const convertStruct toUpper[] = {
   {0x3f2, 0x3f2, -1, 7},
+  {0x3f3, 0x3f3, -1, -116},
   {0x3f5, 0x3f5, -1, -96},
@@ -243,3 +263,3 @@ static const convertStruct toUpper[] = {
   {0x4cf, 0x4cf, -1, -15},
-  {0x4d1, 0x525, 2, -1},
+  {0x4d1, 0x52f, 2, -1},
   {0x561, 0x586, 1, -48},
@@ -286,6 +306,7 @@ static const convertStruct toUpper[] = {
   {0x2cec, 0x2cee, 2, -1},
+  {0x2cf3, 0x2cf3, -1, -1},
   {0x2d00, 0x2d25, 1, -7264},
-  {0xa641, 0xa65f, 2, -1},
-  {0xa663, 0xa66d, 2, -1},
-  {0xa681, 0xa697, 2, -1},
+  {0x2d27, 0x2d2d, 6, -7264},
+  {0xa641, 0xa66d, 2, -1},
+  {0xa681, 0xa69b, 2, -1},
   {0xa723, 0xa72f, 2, -1},
@@ -294,5 +315,8 @@ static const convertStruct toUpper[] = {
   {0xa77f, 0xa787, 2, -1},
-  {0xa78c, 0xa78c, -1, -1},
+  {0xa78c, 0xa791, 5, -1},
+  {0xa793, 0xa797, 4, -1},
+  {0xa799, 0xa7a9, 2, -1},
   {0xff41, 0xff5a, 1, -32},
   {0x10428, 0x1044f, 1, -40},
+  {0x118c0, 0x118df, 1, -32},
 };
@@ -307,6 +331,6 @@ static const struct interval combining[]
   {0x0610, 0x061a},
-  {0x064b, 0x065e},
+  {0x064b, 0x065f},
   {0x0670, 0x0670},
   {0x06d6, 0x06dc},
-  {0x06de, 0x06e4},
+  {0x06df, 0x06e4},
   {0x06e7, 0x06e8},
@@ -321,6 +345,7 @@ static const struct interval combining[]
   {0x0829, 0x082d},
-  {0x0900, 0x0903},
-  {0x093c, 0x093c},
-  {0x093e, 0x094e},
-  {0x0951, 0x0955},
+  {0x0859, 0x085b},
+  {0x08e4, 0x0903},
+  {0x093a, 0x093c},
+  {0x093e, 0x094f},
+  {0x0951, 0x0957},
   {0x0962, 0x0963},
@@ -359,3 +384,3 @@ static const struct interval combining[]
   {0x0bd7, 0x0bd7},
-  {0x0c01, 0x0c03},
+  {0x0c00, 0x0c03},
   {0x0c3e, 0x0c44},
@@ -365,3 +390,3 @@ static const struct interval combining[]
   {0x0c62, 0x0c63},
-  {0x0c82, 0x0c83},
+  {0x0c81, 0x0c83},
   {0x0cbc, 0x0cbc},
@@ -372,3 +397,3 @@ static const struct interval combining[]
   {0x0ce2, 0x0ce3},
-  {0x0d02, 0x0d03},
+  {0x0d01, 0x0d03},
   {0x0d3e, 0x0d44},
@@ -398,3 +423,3 @@ static const struct interval combining[]
   {0x0f86, 0x0f87},
-  {0x0f90, 0x0f97},
+  {0x0f8d, 0x0f97},
   {0x0f99, 0x0fbc},
@@ -410,3 +435,3 @@ static const struct interval combining[]
   {0x109a, 0x109d},
-  {0x135f, 0x135f},
+  {0x135d, 0x135f},
   {0x1712, 0x1714},
@@ -415,3 +440,3 @@ static const struct interval combining[]
   {0x1772, 0x1773},
-  {0x17b6, 0x17d3},
+  {0x17b4, 0x17d3},
   {0x17dd, 0x17dd},
@@ -427,2 +452,3 @@ static const struct interval combining[]
   {0x1a7f, 0x1a7f},
+  {0x1ab0, 0x1abe},
   {0x1b00, 0x1b04},
@@ -431,3 +457,4 @@ static const struct interval combining[]
   {0x1b80, 0x1b82},
-  {0x1ba1, 0x1baa},
+  {0x1ba1, 0x1bad},
+  {0x1be6, 0x1bf3},
   {0x1c24, 0x1c37},
@@ -436,7 +463,9 @@ static const struct interval combining[]
   {0x1ced, 0x1ced},
-  {0x1cf2, 0x1cf2},
-  {0x1dc0, 0x1de6},
-  {0x1dfd, 0x1dff},
+  {0x1cf2, 0x1cf4},
+  {0x1cf8, 0x1cf9},
+  {0x1dc0, 0x1df5},
+  {0x1dfc, 0x1dff},
   {0x20d0, 0x20f0},
   {0x2cef, 0x2cf1},
+  {0x2d7f, 0x2d7f},
   {0x2de0, 0x2dff},
@@ -445,3 +474,4 @@ static const struct interval combining[]
   {0xa66f, 0xa672},
-  {0xa67c, 0xa67d},
+  {0xa674, 0xa67d},
+  {0xa69f, 0xa69f},
   {0xa6f0, 0xa6f1},
@@ -458,2 +488,3 @@ static const struct interval combining[]
   {0xa9b3, 0xa9c0},
+  {0xa9e5, 0xa9e5},
   {0xaa29, 0xaa36},
@@ -461,3 +492,3 @@ static const struct interval combining[]
   {0xaa4c, 0xaa4d},
-  {0xaa7b, 0xaa7b},
+  {0xaa7b, 0xaa7d},
   {0xaab0, 0xaab0},
@@ -467,2 +498,4 @@ static const struct interval combining[]
   {0xaac1, 0xaac1},
+  {0xaaeb, 0xaaef},
+  {0xaaf5, 0xaaf6},
   {0xabe3, 0xabea},
@@ -471,4 +504,6 @@ static const struct interval combining[]
   {0xfe00, 0xfe0f},
-  {0xfe20, 0xfe26},
+  {0xfe20, 0xfe2d},
   {0x101fd, 0x101fd},
+  {0x102e0, 0x102e0},
+  {0x10376, 0x1037a},
   {0x10a01, 0x10a03},
@@ -478,4 +513,33 @@ static const struct interval combining[]
   {0x10a3f, 0x10a3f},
-  {0x11080, 0x11082},
+  {0x10ae5, 0x10ae6},
+  {0x11000, 0x11002},
+  {0x11038, 0x11046},
+  {0x1107f, 0x11082},
   {0x110b0, 0x110ba},
+  {0x11100, 0x11102},
+  {0x11127, 0x11134},
+  {0x11173, 0x11173},
+  {0x11180, 0x11182},
+  {0x111b3, 0x111c0},
+  {0x1122c, 0x11237},
+  {0x112df, 0x112ea},
+  {0x11301, 0x11303},
+  {0x1133c, 0x1133c},
+  {0x1133e, 0x11344},
+  {0x11347, 0x11348},
+  {0x1134b, 0x1134d},
+  {0x11357, 0x11357},
+  {0x11362, 0x11363},
+  {0x11366, 0x1136c},
+  {0x11370, 0x11374},
+  {0x114b0, 0x114c3},
+  {0x115af, 0x115b5},
+  {0x115b8, 0x115c0},
+  {0x11630, 0x11640},
+  {0x116ab, 0x116b7},
+  {0x16af0, 0x16af4},
+  {0x16b30, 0x16b36},
+  {0x16f51, 0x16f7e},
+  {0x16f8f, 0x16f92},
+  {0x1bc9d, 0x1bc9e},
   {0x1d165, 0x1d169},
@@ -486,2 +550,3 @@ static const struct interval combining[]
   {0x1d242, 0x1d244},
+  {0x1e8d0, 0x1e8d6},
   {0xe0100, 0xe01ef},
@@ -555,2 +620,3 @@ static const convertStruct foldCase[] =
   {0x376, 0x376, -1, 1},
+  {0x37f, 0x37f, -1, 116},
   {0x386, 0x386, -1, 38},
@@ -582,5 +648,6 @@ static const convertStruct foldCase[] =
   {0x4c1, 0x4cd, 2, 1},
-  {0x4d0, 0x524, 2, 1},
+  {0x4d0, 0x52e, 2, 1},
   {0x531, 0x556, 1, 48},
   {0x10a0, 0x10c5, 1, 7264},
+  {0x10c7, 0x10cd, 6, 7264},
   {0x1e00, 0x1e94, 2, 1},
@@ -634,5 +701,5 @@ static const convertStruct foldCase[] =
   {0x2ceb, 0x2ced, 2, 1},
-  {0xa640, 0xa65e, 2, 1},
-  {0xa662, 0xa66c, 2, 1},
-  {0xa680, 0xa696, 2, 1},
+  {0x2cf2, 0xa640, 31054, 1},
+  {0xa642, 0xa66c, 2, 1},
+  {0xa680, 0xa69a, 2, 1},
   {0xa722, 0xa72e, 2, 1},
@@ -643,4 +710,14 @@ static const convertStruct foldCase[] =
   {0xa78b, 0xa78b, -1, 1},
+  {0xa78d, 0xa78d, -1, -42280},
+  {0xa790, 0xa792, 2, 1},
+  {0xa796, 0xa7a8, 2, 1},
+  {0xa7aa, 0xa7aa, -1, -42308},
+  {0xa7ab, 0xa7ab, -1, -42319},
+  {0xa7ac, 0xa7ac, -1, -42315},
+  {0xa7ad, 0xa7ad, -1, -42305},
+  {0xa7b0, 0xa7b0, -1, -42258},
+  {0xa7b1, 0xa7b1, -1, -42282},
   {0xff21, 0xff3a, 1, 32},
   {0x10400, 0x10427, 1, 40},
+  {0x118a0, 0x118bf, 1, 32},
 };
@@ -648,4 +725,2 @@ static const struct interval doublewidth
   {0x1100, 0x115f},
-  {0x11a3, 0x11a7},
-  {0x11fa, 0x11ff},
   {0x2329, 0x232a},
@@ -655,9 +730,8 @@ static const struct interval doublewidth
   {0x2ff0, 0x2ffb},
-  {0x3000, 0x3029},
-  {0x3030, 0x303e},
+  {0x3000, 0x303e},
   {0x3041, 0x3096},
-  {0x309b, 0x30ff},
+  {0x3099, 0x30ff},
   {0x3105, 0x312d},
   {0x3131, 0x318e},
-  {0x3190, 0x31b7},
+  {0x3190, 0x31ba},
   {0x31c0, 0x31e3},
@@ -671,4 +745,2 @@ static const struct interval doublewidth
   {0xac00, 0xd7a3},
-  {0xd7b0, 0xd7c6},
-  {0xd7cb, 0xd7fb},
   {0xf900, 0xfaff},
@@ -680,5 +752,7 @@ static const struct interval doublewidth
   {0xffe0, 0xffe6},
-  {0x1f200, 0x1f200},
-  {0x1f210, 0x1f231},
+  {0x1b000, 0x1b001},
+  {0x1f200, 0x1f202},
+  {0x1f210, 0x1f23a},
   {0x1f240, 0x1f248},
+  {0x1f250, 0x1f251},
   {0x20000, 0x2fffd},
@@ -740,2 +814,3 @@ static const struct interval ambiguous[]
   {0x02df, 0x02df},
+  {0x0300, 0x036f},
   {0x0391, 0x03a1},
@@ -850,2 +925,3 @@ static const struct interval ambiguous[]
   {0xe000, 0xf8ff},
+  {0xfe00, 0xfe0f},
   {0xfffd, 0xfffd},
@@ -853,15 +929,5 @@ static const struct interval ambiguous[]
   {0x1f110, 0x1f12d},
-  {0x1f131, 0x1f131},
-  {0x1f13d, 0x1f13d},
-  {0x1f13f, 0x1f13f},
-  {0x1f142, 0x1f142},
-  {0x1f146, 0x1f146},
-  {0x1f14a, 0x1f14e},
-  {0x1f157, 0x1f157},
-  {0x1f15f, 0x1f15f},
-  {0x1f179, 0x1f179},
-  {0x1f17b, 0x1f17c},
-  {0x1f17f, 0x1f17f},
-  {0x1f18a, 0x1f18d},
-  {0x1f190, 0x1f190},
+  {0x1f130, 0x1f169},
+  {0x1f170, 0x1f19a},
+  {0xe0100, 0xe01ef},
   {0xf0000, 0xffffd},
```

ZyX-I committed:
Generate unicode tables with script when building

justinmk commented:
Very cool!

tarruda commented:
:+1: great stuff

justinmk commented on scripts/genunicodetables.lua:
```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/lua
```
Couldn't hurt to make this available in `runtime/tools` like the old one.

ZyX-I replied:
What for? Other generator scripts are not there, VimL is no longer used. And former `runtime/tools/unicode.vim` was creating a number of buffers in whatever directory it was run (it was needed to buffer contents to `mbyte.c` manually then) and new script is not going to do anything like this (it does not run Vim in first place).

justinmk replied:
Fair enough

Pyrohh replied:
Also regarding this line: IIRC FreeBSD installs lua to `/usr/local/bin`, so this should be `#!/usr/bin/env lua`.

ZyX-I replied:
@Pyrohh Other scripts are using `#!/usr/bin/lua` or nothing. In any case they are run by directly calling lua executable and not by calling scripts themselves, so this is irrelevant.

jszakmeister replied:
I don't care either way, but I think if we're arguing it doesn't matter what the shebang is because we run it via lua directly, then it doesn't make sense for it to be there at all and for the file to be marked executable.  FWIW, there's only one other script, and it could be changed too (not as part of this PR). :-)

I don't feel strongly either way about it, just pointing out where that line of thinking leads. :-)

ZyX-I replied:
You are right. I removed shebang and +x here.

ZyX-I commented:
About `file(DOWNLOAD`: debian maintainers definitely are not going to like this due to https://wiki.debian.org/ReproducibleBuilds.

Possible solutions:

1. Check if there are files `unicode/…` in the repository (intended to be put there by package maintainers). If there are, copy them to `build/unicode/…`, otherwise download.
2. Put necessary files into the repository and periodically redownload them by hand.
3. Like 1., but make download happen when running `make` (I mean, make in the *build* directory; I never use Makefile in the repository root). Not sure how to make this platform-independent.

justinmk commented:
> About file(DOWNLOAD: debian maintainers definitely are not going to like this 

I'm also not very comfortable having the source tree contents change based on a definition that lives on a website. I think we should check the generated file into source control, and have a bot-ci script alert us whenever `http://unicode.org/Public/UNIDATA/UnicodeData.txt` and friends have changed. Thoughts?

ZyX-I commented:
This is variant 2. I am personally OK with this:

1. Download, put and commit those files into `/unicode` directory.
2. Add a script which does 1. for you.
3. Remove `file(DOWNLOAD` from CMake.

ZyX-I committed:
Create script that downloads and commits UNIDATA files

ZyX-I committed:
Update unicode files

fwalch commented:
As we now save a copy of the unicode files, their [license (bottom of page)](http://www.unicode.org/copyright.html) should probably be added to the `unicode/` directory as well:

```
(a) this copyright and permission notice appear with all copies 
of the Data Files or Software,
```

justinmk commented:
Does the dirgraphs help file need to be updated?

ZyX-I commented:
@justinmk How is digraphs help file related?

Oberon00 commented on scripts/download-unicode-files.sh:
```diff
@@ -0,0 +1,35 @@
+
+(
+  cd "$UNIDIR"
+  git commit -m "Update unicode files" -- *.txt
```
This is somewhat dangerous, since other files could have been `git add`ded by the user before.

ZyX-I replied:
@Oberon00 User `git add`ed `*.txt` file in `unicode` directory? It is his problem. Files in other locations will not be committed.

ZyX-I replied:
Though I already have a nice array of files to be committed. Can just use this.

Oberon00 replied:
Ah sorry, i missed the `-- *.txt`. Yes then this should be really unlikely to cause any problems ;).

justinmk commented:
@ZyX-I I noticed the descriptions like `LATIN CAPITAL LETTER A` are same as in the digraph file, and thought maybe if the unicode.org files are adding or changing descriptions, then the descriptions in digraph help file may need to be updated.

ZyX-I commented:
@justinmk Never heard them changing descriptions. New characters with new descriptions are irrelevant.

justinmk commented:
Alrighty then.

justinmk commented on unicode/CaseFolding.txt:
```diff
@@ -0,0 +1,1273 @@
+# CaseFolding-7.0.0.txt
```
`third-party/unicode/` ?

justinmk commented:
Perhaps just commit the generated file, not the original files?

ZyX-I commented:
I don’t want to leave precedent of committing generated file that was not meant to be editable (some of the current files like `ex_cmds.lua` are initially generated, but intended to be editable by hand since than). Also no files from `third-party` are ever used in the build process (unless you use top-level Makefile): so far this directory was only meant for the stuff necessary for NeoVim dependencies, but not for NeoVim itself.

Also these files from unicode.org are going to be useful should we finally add support for something like Perl `\p` to the regex engine.

jszakmeister commented on scripts/download-unicode-files.sh:
```diff
@@ -0,0 +1,35 @@
+  (
+    cd "$UNIDIR"
+    git add $filename
+  )
```
You can probably just remove this step (the cd and add)...

jszakmeister commented on scripts/download-unicode-files.sh:
```diff
@@ -0,0 +1,35 @@
+
+(
+  cd "$UNIDIR"
+  git commit -m "Update unicode files" -- $files
```
Because this essentially does it (as long as it's under version control already).

ZyX-I replied:
The last commit in sequence was done by the script. And that was the initial intention behind `git add` above.

glts commented:
Note: this will take care of [Vim patch 7.4.575](https://github.com/vim/vim/commit/v7-4-575).

ZyX-I commented:
Wondering whether anybody is going to merge this. I have rebased and updated to unicode-8.0 (differences with #2835 are there because when this PR was created latest version was 7.0).

tarruda commented:
> Wondering whether anybody is going to merge this. I have rebased and updated to unicode-8.0.0 (differences with #2835 are there because when this PR was created latest version was 7.0.0).

In any case, can you squash the last two commits? I don't think it would be useful to keep these differences in our history

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -2188,192 +1735,11 @@ int utf_class(int c)
- * They must be in numeric order, because we use binary search.
- * An entry such as {0x41,0x5a,1,32} means that Unicode characters in the
- * range from 0x41 to 0x5a inclusive, stepping by 1, are changed to
- * folded/upper/lower by adding 32.
```
Can this comment be added to the generated tables, so the meaning of these tables is not "lost"?

ZyX-I replied:
I have added explanations to the top of the generator script.

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -944,225 +945,6 @@ static bool intable(const struct interval *table, size_t n_items, int c)
 int utf_char2cells(int c)
 {
-  /* Sorted list of non-overlapping intervals of East Asian double width
-   * characters, generated with ../runtime/tools/unicode.vim. */
```
Preserve comment?

oni-link commented on src/nvim/mbyte.c:
```diff
@@ -1806,241 +1588,6 @@ int utf_char2bytes(int c, char_u *buf)
 bool utf_iscomposing(int c)
 {
-  /* Sorted list of non-overlapping intervals.
-   * Generated by ../runtime/tools/unicode.vim. */
```
Preserve comment?

oni-link commented:
Beside perhaps adding some comments to the generated tables: `scripts/genunicodetables.lua` LGTM.
Can't say anything about the CMake and shell script.

ZyX-I committed:
documentation: Document table meanings in comment to genunicodetables

ZyX-I commented:
I have no idea what does the QuickBuild failure mean:

```
09:37:12,860 ERROR - Step 'master>consolidate-and-publish>consolidate-junit-xml-reports' is failed: org.xml.sax.SAXParseException; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
    at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1239)
    at com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:649)
    at groovy.util.XmlParser.parse(XmlParser.java:175)
    at groovy.util.XmlParser.parseText(XmlParser.java:208)
    at groovy.util.XmlParser$parseText.call(Unknown Source)
    at script14368755641011594417190.combineReportsInto(script14368755641011594417190.groovy:16)
    at script14368755641011594417190$combineReportsInto.callCurrent(Unknown Source)
    at script14368755641011594417190.run(script14368755641011594417190.groovy:42)
    at com.pmease.quickbuild.plugin.basis.BasisPlugin$28.evaluate(BasisPlugin.java:348)
    at com.pmease.quickbuild.DefaultScriptEngine.evaluate(DefaultScriptEngine.java:81)
    at com.pmease.quickbuild.DefaultScriptEngine.interpolate(DefaultScriptEngine.java:105)
    at com.pmease.quickbuild.DefaultScriptEngine$Interpolator.intercept(DefaultScriptEngine.java:281)
    at com.pmease.quickbuild.plugin.basis.ScriptStep$$EnhancerByCGLIB$$61d866d7.getScript(<generated>)
    at com.pmease.quickbuild.plugin.basis.ScriptStep.run(ScriptStep.java:47)
    at com.pmease.quickbuild.plugin.basis.ScriptStep$$EnhancerByCGLIB$$61d866d7.CGLIB$run$1(<generated>)
    at com.pmease.quickbuild.plugin.basis.ScriptStep$$EnhancerByCGLIB$$61d866d7$$FastClassByCGLIB$$b862b47d.invoke(<generated>)
    at net.sf.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:215)
    at com.pmease.quickbuild.DefaultScriptEngine$Interpolator.intercept(DefaultScriptEngine.java:273)
    at com.pmease.quickbuild.plugin.basis.ScriptStep$$EnhancerByCGLIB$$61d866d7.run(<generated>)
    at com.pmease.quickbuild.stepsupport.Step.execute(Step.java:539)
    at com.pmease.quickbuild.stepsupport.StepExecutionJob.executeStepAwareJob(StepExecutionJob.java:30)
    at com.pmease.quickbuild.stepsupport.StepAwareJob.executeBuildAwareJob(StepAwareJob.java:45)
    at com.pmease.quickbuild.BuildAwareJob.execute(BuildAwareJob.java:60)
    at com.pmease.quickbuild.grid.GridJob.run(GridJob.java:106)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

ZyX-I commented:
Rebased due to conflicts with #2288.

tarruda commented:
> I have no idea what does the QuickBuild failure mean:

I have skimmed through the logs and couldn't find anything either, @jszakmeister can you spot what's wrong?

jszakmeister commented:
@ZyX-I @tarruda It means that crap from stdout was squirted into the junit XML file making the XML file invalid.  I really wish busted provided an option to point to a file, and would emit the junit report to that instead of using stdout this way.

ZyX-I commented:
@jszakmeister I would suggest to file this request to busted developers. Do not think they will find it unreasonable.

ZyX-I commented:
@tarruda @justinmk @oni-link Any objections to merging this?

oni-link commented:
Comments added, so LGTM.

justinmk commented:
@ZyX-I So this doesn't happen at build time, correct? We must run `scripts/download-unicode-files.sh` when we want to update the unicode definitions. I will note this in my `develop.txt` fork.

jamessan commented:
Why not allow specifying an external location for the data files?  Going back to @ZyX-I's comment on Debian, there's a unicode-data package that a neovim package could Build-Depend on and then just point neovim's build to its files.

justinmk commented:
We include the Unicode data in our source. It is not downloaded at compile time, shouldn't Debian be OK with this? Neovim should be consistent across platforms and distros. If something needs to change it should be patched in neovim upstream, not in a distro package.

jamessan commented:
I was using Debian as an example, but the question was general.  Why not allow specifying an external data source?  It's similar to not bundling every single library you use, because then I have to wait for you to update them instead of just using the upstream sources.  Similarly, if your dependencies need patching, they should be patched in their respective upstream, not in neovim.

justinmk commented:
As you know, these definitions are hard-coded in the Vim source. Your have a good point though... I have no objections.
